gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java

```diff
@@ -52,6 +54,21 @@
 
   public static final String HDFS_ILLEGAL_TOKEN_REGEX = "[\\s:\\\\]";
 
+  /**
+   * A {@link Collection} of all S3 schemes.
+   *
+   * <p>
+   *   The following important properties are useful to remember when writing code that is compatible with S3:
+   *   <ul>
+   *     <li>Renames are not atomic, and require copying the entire source file to the destination file</li>
+   *     <li>Writes to S3 will first go to the local filesystem, when the stream is closed the local file will be
+   *     uploaded to S3</li>
+   *   </ul>
+   * </p>
+   */
+  public static final Collection<String> S3_SCHEMES = ImmutableSortedSet.orderedBy(
```

Comment:
One issue here is that I think what we are really trying to identify are schemes which support atomic move/copy.  

This reminds me of a conversation with @ibuenros:
-- ibuenros
...
Maybe we could have a FileSystem decorator instead with a few of the methods you mention. We could then detect certain known schemes with incompatibilities and apply fixes across the entire code, but any FileSystem we haven't considered yet would still work appropriately.
  
-- kadaan
That would be great, but I think users of gobblin should be able to provide new decorators without having to extend gobblin itself.